api/pkg/transformer/types/encoder/cyclical_encoder.go

```diff
@@ -0,0 +1,312 @@
+			encodedCos = append(encodedCos, math.Cos(phase))
+			encodedSin = append(encodedSin, math.Sin(phase))
+		}
+	}
+
+	return map[string]interface{}{
+		column + "_x": encodedCos,
+		column + "_y": encodedSin,
+	}, nil
+}
+
+//Computes the number of seconds past the beginning of a pre-defined cycle
+//Only works with PeriodType with variable cycle time such as Month, Year etc
+//For period type with fixed cycle time, it is handled differently by encoder and
+//does not need the cycle time to be computed
+func getCycleTime(periodType interface{}, t time.Time) (float64, error) {
+	_, pType := periodType.(spec.PeriodType)
+	if !pType {
+		return 0, fmt.Errorf("invalid type for periodType")
+	}
+
+	switch periodType {
+	case spec.PeriodType_MONTH:
+		dayElapsed := t.Day() - 1
+		hr, min, sec := t.Clock()
+		elapsed := float64(dayElapsed*DayInSec + hr*HourInSec + min*MinInSec + sec)
+		return elapsed, nil
+	case spec.PeriodType_QUARTER:
+		dayElapsed := t.YearDay() - 1
+		hr, min, sec := t.Clock()
+		elapsed := float64(dayElapsed*DayInSec + hr*HourInSec + min*MinInSec + sec)
+
+		if t.Month() <= 3 {
+			return elapsed, nil
+		} else if t.Month() <= 6 {
+			if isLeapYear(t.Year()) {
+				return elapsed - Q1LeapInSec, nil
+			}
+			return elapsed - Q1InSec, nil
+		} else if t.Month() <= 9 {
+			if isLeapYear(t.Year()) {
+				return elapsed - H1LeapInSec, nil
+			}
+			return elapsed - H1InSec, nil
+		}
+
+		if isLeapYear(t.Year()) {
+			return elapsed - H1LeapInSec - Q3InSec, nil
+		}
+		return elapsed - H1InSec - Q3InSec, nil
+
+	case spec.PeriodType_HALF:
+		dayElapsed := t.YearDay() - 1
+		hr, min, sec := t.Clock()
+		elapsed := float64(dayElapsed*DayInSec + hr*HourInSec + min*MinInSec + sec)
+
+		if t.Month() <= 6 {
+			return elapsed, nil
+		}
+
+		if isLeapYear(t.Year()) {
+			return elapsed - H1LeapInSec, nil
+		}
+		return elapsed - H1InSec, nil
+
+	case spec.PeriodType_YEAR:
+		dayElapsed := t.YearDay() - 1
+		hr, min, sec := t.Clock()
+		elapsed := float64(dayElapsed*DayInSec + hr*HourInSec + min*MinInSec + sec)
+		return elapsed, nil
+	}
+
+	return 0, fmt.Errorf("period type is undefined for this use case")
+}
+
+//Computes the angle in radians represented by per unit second of a pre-defined period
+//This is derived from the formula for calculating phase:
+//phase = time passed / period * 2pi
+//By rearranging the formula (for optimizing computation) into this:
+//phase = time pass * 2pi / period
+//we define unit angle as (2pi / period)
+//The motivation is that we can pre-compute this value once and use it repeatedly.
+func getUnitAngle(periodType interface{}, t time.Time) (float64, error) {
+	_, pType := periodType.(spec.PeriodType)
+	if !pType {
+		return 0, fmt.Errorf("invalid type for periodType")
+	}
+
+	switch periodType {
+	case spec.PeriodType_MONTH:
+		if t.Month() == 2 && isLeapYear(t.Year()) {
+			return MonthInSec[13], nil
+		}
+		return MonthInSec[int(t.Month())], nil
+	case spec.PeriodType_QUARTER:
+		if t.Month() <= 3 {
```

Comment:
better to store value 3, 6, 9 into variable with proper naming so other understand the meaning of that value